File: src/com/maddyhome/idea/vim/helper/MacKeyRepeat.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2021 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.maddyhome.idea.vim.helper;

import com.google.common.io.CharStreams;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

/**
* @author vlan
*/
public class MacKeyRepeat {
  @VimNlsSafe public static final String FMT = "defaults %s -globalDomain ApplePressAndHoldEnabled";
  @NotNull private static final MacKeyRepeat INSTANCE = new MacKeyRepeat();
  @NonNls private static final String EXEC_COMMAND = "launchctl stop com.apple.SystemUIServer.agent";
  @NonNls private static final String delete = "delete";
  @NonNls private static final String write = "write";
  @NonNls private static final String read = "read";

  public static @NotNull MacKeyRepeat getInstance() {
    return INSTANCE;
  }

  public @Nullable Boolean isEnabled() {
    final String command = String.format(FMT, read);
    try {
      final Process process = Runtime.getRuntime().exec(command);
      final String data = read(process.getInputStream()).trim();
      try {
        return Integer.parseInt(data) == 0;
      } catch (NumberFormatException e) {
        return null;
      }
    }
    catch (IOException e) {
      return null;
    }
  }

  public void setEnabled(@Nullable Boolean value) {
    final String command;
    if (value == null) {
      command = String.format(FMT, delete);
    }
    else {
      final String arg = value ? "0" : "1";
      command = String.format(FMT, write) + " " + arg;
    }
    try {
      final Runtime runtime = Runtime.getRuntime();
      final Process defaults = runtime.exec(command);
      defaults.waitFor();
      final Process restartSystemUI = runtime.exec(EXEC_COMMAND);
      restartSystemUI.waitFor();
    }
    catch (IOException | InterruptedException ignored) {
    }
  }

  private static @NotNull String read(@NotNull InputStream stream) throws IOException {
    return CharStreams.toString(new InputStreamReader(stream));
  }
}

File: src/main/java/com/maddyhome/idea/vim/helper/MacKeyRepeat.kt
/*
 * Copyright 2003-2023 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */
package com.maddyhome.idea.vim.helper

import com.google.common.io.CharStreams
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader

object MacKeyRepeat {
  var isEnabled: Boolean?
    get() {
      return try {
        val process = Runtime.getRuntime().exec(READ_COMMAND)
        val data = read(process.inputStream).trim().toIntOrNull() ?: return null
        data == 0
      } catch (_: IOException) {
        null
      }
    }
    set(value) {
      val command: Array<String>
      if (value == null) {
        command = DELETE_COMMAND
      } else {
        val arg = if (value) "0" else "1"
        command = WRITE_COMMAND + arg
      }
      try {
        val runtime = Runtime.getRuntime()
        val defaults = runtime.exec(command)
        defaults.waitFor()
        val restartSystemUI: Process = runtime.exec(EXEC_COMMAND)
        restartSystemUI.waitFor()
      } catch (_: IOException) {
      } catch (_: InterruptedException) {
      }
    }

  private val EXEC_COMMAND = arrayOf("launchctl", "stop", "com.apple.SystemUIServer.agent")
  private val READ_COMMAND = arrayOf("defaults", "read", "-globalDomain", "ApplePressAndHoldEnabled")
  private val WRITE_COMMAND = arrayOf("defaults", "write", "-globalDomain", "ApplePressAndHoldEnabled")
  private val DELETE_COMMAND = arrayOf("defaults", "delete", "-globalDomain", "ApplePressAndHoldEnabled")

  @Throws(IOException::class)
  private fun read(stream: InputStream): String {
    return CharStreams.toString(InputStreamReader(stream))
  }
}
